core/src/main/java/org/elasticsearch/cockroach/CockroachActionExecutor.java

```diff
@@ -0,0 +1,363 @@
+        private final RunningCockroachTask<Response> task;
+
+        public CockroachTaskListener(final RunningCockroachTask<Response> task) {
+            this.task = task;
+        }
+
+        @Override
+        public void onResponse(Response response) {
+            startNotification(task, response);
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            startNotification(task, e);
+        }
+    }
+
+    private class PublishingFailureListener<Response extends CockroachResponse> implements ActionListener<Empty> {
+        private final RunningCockroachTask<Response> task;
+
+        public PublishingFailureListener(final RunningCockroachTask<Response> task) {
+            this.task = task;
+        }
+
+
+        @Override
+        public void onResponse(Empty empty) {
+            logger.trace("notification for task {} was successful", task.getUuid());
+            if (task.markAsNotified() == false) {
+                logger.warn("attempt to mark task {} in the {} state as NOTIFIED", task.getUuid(), task.getState());
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            logger.debug("notification for task {} failed - retrying", e, task.getUuid());
+            if (task.notificationFailed() == false) {
+                logger.warn("attempt to mark restart notification for task {} in the {} state", task.getUuid(), task.getState());
+            }
+        }
+    }
+
+    public enum State {
+        STARTED,  // the task is currently running
+        DONE,     // the task is done running and trying to notify caller
+        FAILED_NOTIFICATION, // the caller notification failed
+        NOTIFIED // the caller was notified, the task can be removed
+    }
+
+    private static class RunningCockroachTask<Response extends CockroachResponse> implements Provider<Task.Status> {
+        private final Task task;
+        private final String uuid;
+        private final AtomicReference<State> state;
+        @Nullable
+        private Response response;
+        @Nullable
+        private Exception failure;
+
+        public RunningCockroachTask(Task task, String uuid) {
+            this(task, uuid, State.STARTED);
+        }
+
+        public RunningCockroachTask(Task task, String uuid, State state) {
+            this.task = task;
+            this.uuid = uuid;
+            this.state = new AtomicReference<>(state);
+        }
+
+        public Task getTask() {
+            return task;
+        }
+
+        public String getUuid() {
+            return uuid;
+        }
+
+        public State getState() {
+            return state.get();
+        }
+
+        public Response getResponse() {
+            return response;
+        }
+
+        public Exception getFailure() {
+            return failure;
+        }
+
+        public boolean startNotification(Response response) {
```

Comment:
This method name confused me until I traced through what it meant, I think a better name would be `startCompletionNotification` or something similar.  Also this method and the other methods that change state could use javadocs to explain the state transitions.